src/async.rs

```diff
@@ -0,0 +1,228 @@
+//! Provides asynchronous retry functionality via `futures`.
+//!
+//! # Examples
+//!
+//! ```rust
+//! # extern crate futures;
+//! # extern crate tokio_timer;
+//! # extern crate retry;
+//! # use std::default::Default;
+//! # use futures::future::Future;
+//! use tokio_timer::Timer;
+//! use retry::delay::{Exponential, jitter};
+//! use retry::async::retry;
+//!
+//! pub fn main() {
+//!     let timer = Timer::default();
+//!     let strategy = Exponential::from_millis(10)
+//!         .map(jitter).take(3);
+//!
+//!     let mut collection = vec![1, 2, 3].into_iter();
+//!
+//!     let future = retry(timer, strategy, || match collection.next() {
+//!         Some(n) if n == 3 => Ok("n is 3!"),
+//!         Some(_) => Err("n must be 3!"),
+//!         None => Err("n was never 3!"),
+//!     });
+//!
+//!     let result = future.wait();
+//!
+//!     assert!(result.is_ok());
+//! }
+//! ```
+
+use either::Either;
```

Comment:
I like that style a lot. Interestingly, I tend to follow it in other programming languages, but for some reason not yet in Rust. Time to change that :)